## AIVLE School [AI 미니프로젝트] '서울시 생활정보 기반 대중교통 수요 분석'<br>

---

# 2. 구 별 이동인구 분석
### 서울시 생활이동 정보를 기반으로 서울시 자치구 별 이동인구를 분석합니다

In [1]:
# 필요 라이브러리를 설치합니다.
%pip install pandas seaborn

## 데이터 안내

[데이터 파일명]
* 2. seoul_moving.csv

[데이터 정보]
* 서울 시 구별 이동 2022년 4월 데이터

[데이터 컬럼 정보]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만(0~3)의 경우 * 처리

---

# 2.1.데이터 불러오기

#### [2.1.1] 데이터 로딩
* Pandas 라이브러리를 활용해서 '2. seoul_moving.csv'파일을 'seoul_moving' 변수에 저장하세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요
        * index column은 0번 열 입니다
    * 데이터를 불러올 때 오류가 발생한다면, 오류에 대해 직접 검색하여 해결해보세요.

In [2]:
import sys
import numpy as np
import pandas as pd

In [3]:
# 아래에 코드를 작성하고 결과를 확인합니다.
path='2.seoul_moving.csv'
seoul_moving = pd.read_csv('2. seoul_moving.csv',encoding='utf-8')

In [4]:
# 데이터 프레임의 Shape을 확인합니다.
seoul_moving.shape

(28923054, 11)

---

# 2.2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [2.2.1] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head', 'tail' 등 전부 활용해 보세요.

In [5]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving.describe()

Unnamed: 0        대상연월          도착시간     출발 시군구 코드     도착 시군구 코드  \
count  2.892305e+07  28923054.0  2.892305e+07  2.892305e+07  2.892305e+07   
mean   7.141390e+05    202204.0  1.365597e+01  1.712362e+04  1.719750e+04   
std    4.753792e+05         0.0  5.849675e+00  9.046478e+03  9.063818e+03   
min    0.000000e+00    202204.0  0.000000e+00  1.101000e+04  1.101000e+04   
25%    3.012810e+05    202204.0  1.000000e+01  1.110000e+04  1.111000e+04   
50%    6.556670e+05    202204.0  1.400000e+01  1.120000e+04  1.120000e+04   
75%    1.082991e+06    202204.0  1.800000e+01  3.101400e+04  3.102100e+04   
max    1.871922e+06    202204.0  2.300000e+01  3.900000e+04  3.900000e+04   

                 나이   평균 이동 시간(분)  
count  2.892305e+07  2.892305e+07  
mean   4.279614e+01  3.279829e+01  
std    1.777485e+01  2.584213e+01  
min    0.000000e+00  1.000000e+01  
25%    3.000000e+01  2.000000e+01  
50%    4.500000e+01  3.000000e+01  
75%    5.500000e+01  4.000000e+01  
max    8.000000e+01  7.700000e+02

In [6]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28923054 entries, 0 to 28923053
Data columns (total 11 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   Unnamed: 0   int64 
 1   대상연월         int64 
 2   요일           object
 3   도착시간         int64 
 4   출발 시군구 코드    int64 
 5   도착 시군구 코드    int64 
 6   성별           object
 7   나이           int64 
 8   이동유형         object
 9   평균 이동 시간(분)  int64 
 10  이동인구(합)      object
dtypes: int64(7), object(4)
memory usage: 2.4+ GB


In [7]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving.head(5)

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0           0  202204  일     0      11010      11010  F   0   EE           10   
1           1  202204  일     0      11010      11010  F   0   EH           10   
2           2  202204  일     0      11010      11010  F   0   HE           10   
3           3  202204  일     0      11010      11010  F  10   HW           10   
4           4  202204  일     0      11010      11010  F  10   EH           90   

  이동인구(합)  
0   28.02  
1   27.52  
2   27.83  
3    6.05  
4    5.95

In [8]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving.tail(5)

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
28923049      968458  202204  토    23      39000      11250  M  45   EH   
28923050      968459  202204  토    23      39000      11250  M  50   EH   
28923051      968460  202204  토    23      39000      11250  M  55   EE   
28923052      968461  202204  토    23      39000      11250  M  55   EH   
28923053      968462  202204  토    23      39000      11250  M  75   EH   

          평균 이동 시간(분) 이동인구(합)  
28923049           50    3.16  
28923050           70    3.27  
28923051           50    3.31  
28923052           50    3.28  
28923053           30    5.94

* [참고] 1번 열의 의미
    * " 2022년 4월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 H(야간상주지(집))으로 이동한 0 ~ 10세 여성은 총 27.52명이며 이들의 평균 이동시간은 10분이다

#### [2.2.2] 결측치 처리 (NaN이 아닌 결측치 처리)

* '이동인구(합)' column은 이동 인구의 합산인데 숫자가 아닌 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만 즉, 값이 0~3일 경우 별표 * 처리)

    - 해당 값을 어떻게 처리하는 것이 좋을 지 생각해보아서 각자의 방법으로 변경하세요

    - '인구이동(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [9]:
# 이동인구(합) 컬럼의 값 별로 개수를 세어봅니다.
seoul_moving['이동인구(합)'].value_counts()


*          3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [10]:
# 별표(*)로 표현된 값을 대체합니다.
# 해당 값을 어떻게 처리하는 것이 좋을 지 생각해보아서value_counts 방법으로 변경하세요
#(ex: 최빈값, 평균값도 알 수 없으므로 0~3의 중위값인 1.5로 대체)
seoul_moving['이동인구(합)']=np.where(seoul_moving['이동인구(합)']=='*',1.5,seoul_moving['이동인구(합)'])

In [11]:
# 이동인구(합) 컬럼의 값 별로 개수를 다시 세어봅니다.
seoul_moving['이동인구(합)'].value_counts()

1.5        3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [12]:
# 이제 이동인구(합) column의 데이터 타입을 바꿔주세요
seoul_moving = seoul_moving.astype({'이동인구(합)':'float64'})
seoul_moving.dtypes

Unnamed: 0       int64
대상연월             int64
요일              object
도착시간             int64
출발 시군구 코드        int64
도착 시군구 코드        int64
성별              object
나이               int64
이동유형            object
평균 이동 시간(분)      int64
이동인구(합)        float64
dtype: object

#### [2.2.3] 이동유형, 요일, 성별 변수의 value수 파악

In [13]:
# 이동유형 컬럼의 값 별로 개수를 세어봅니다.
# H(야간상주지), W(주간상주지), E(기타)
seoul_moving['이동유형'].value_counts()

EH    5757073
HE    5520005
EE    5085099
WH    3108026
WE    2809148
HW    2695189
EW    2304535
WW    1026134
HH     617845
Name: 이동유형, dtype: int64

In [14]:
# 요일 컬럼의 값 별로 개수를 세어봅니다.
seoul_moving['요일'].value_counts()

토    4690806
금    4641881
목    4011291
수    3955646
화    3926876
월    3906272
일    3790282
Name: 요일, dtype: int64

* [참고] 2022년 4월의 요일 수는 다음과 같습니다.
    * 일 : 4, 월 : 4, 화 : 4, 수 : 4, 목 : 4, 금 : 5, 토 : 5

In [15]:
# 성별 컬럼의 값 별로 개수를 세어봅니다.
seoul_moving['성별'].value_counts()


M    16355029
F    12568025
Name: 성별, dtype: int64

#### [2.2.4] 심야 이동이 가장 많은 요일 확인

In [16]:
# 아래에 코드를 작성하고 결과를 확인합니다.
# 심야시간은 0:00~ 4:59
seoul_moving.loc[seoul_moving['도착시간'].between(0,4)].groupby('요일')['도착시간'].count().idxmax()
# est= temp.loc[temp['도착시간']<=4].groupby('요일')[['이동인구(합)']].count().idxmax()

'토'

#### [2.2.5] 서울 지역의 데이터만 추출

* 서울시의 데이터만 필요하므로 출발과 도착이 모두 서울인 데이터만 추출하세요.
* 아래 사이트에서 서울시 데이터만 추출하기 위해 필요한 정보를 탐색해보세요. (ex. 자치구 코드 정보)
https://data.seoul.go.kr/dataVisual/seoul/seoulLivingMigration.do

In [17]:
# 출발 시군구 코드, 도착 시군구 코드 활용하여 출발과 도착이 모두 서울인 데이터만 추출해보세요.
start=seoul_moving['출발 시군구 코드'].astype(str)
end=seoul_moving['도착 시군구 코드'].astype(str)
seoul_moving['승차자치구'] = start.str[0:2]
seoul_moving['하차자치구'] = end.str[0:2]
seoul_moving

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                  0  202204  일     0      11010      11010  F   0   EE   
1                  1  202204  일     0      11010      11010  F   0   EH   
2                  2  202204  일     0      11010      11010  F   0   HE   
3                  3  202204  일     0      11010      11010  F  10   HW   
4                  4  202204  일     0      11010      11010  F  10   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28923049      968458  202204  토    23      39000      11250  M  45   EH   
28923050      968459  202204  토    23      39000      11250  M  50   EH   
28923051      968460  202204  토    23      39000      11250  M  55   EE   
28923052      968461  202204  토    23      39000      11250  M  55   EH   
28923053      968462  202204  토    23      39000      11250  M  75   EH   

          평균 이동 시간(분)  이동인구(합) 승차자치구 하차자치구  
0                  10    28.02    11    11  
1                  10    27.52    11    11  
2                  10    27.83    11    11  
3                  10     6.05    11    11  
4                  90     5.95    11    11  
...               ...      ...   ...   ...  
28923049           50     3.16    39    11  
28923050           70     3.27    39    11  
28923051           50     3.31    39    11  
28923052           50     3.28    39    11  
28923053           30     5.94    39    11  

[28923054 rows x 13 columns]

In [18]:
seoul_moving = seoul_moving[(seoul_moving['승차자치구']=='11') & (seoul_moving['하차자치구']=='11')]
seoul_moving

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                  0  202204  일     0      11010      11010  F   0   EE   
1                  1  202204  일     0      11010      11010  F   0   EH   
2                  2  202204  일     0      11010      11010  F   0   HE   
3                  3  202204  일     0      11010      11010  F  10   HW   
4                  4  202204  일     0      11010      11010  F  10   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28875001      920410  202204  토    23      11250      11250  M  80   WH   
28875002      920411  202204  토    23      11250      11250  M  80   EH   
28875003      920412  202204  토    23      11250      11250  M  80   EE   
28875004      920413  202204  토    23      11250      11250  M  80   HE   
28875005      920414  202204  토    23      11250      11250  M  80   WE   

          평균 이동 시간(분)  이동인구(합) 승차자치구 하차자치구  
0                  10    28.02    11    11  
1                  10    27.52    11    11  
2                  10    27.83    11    11  
3                  10     6.05    11    11  
4                  90     5.95    11    11  
...               ...      ...   ...   ...  
28875001           10    85.10    11    11  
28875002           10    28.36    11    11  
28875003           20    37.99    11    11  
28875004           10     9.42    11    11  
28875005           10     9.51    11    11  

[10655895 rows x 13 columns]

#### [2.2.6] 서울지역의 구 별로 groupby하여 데이터를 확인해보세요. (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* [참고] 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [20]:
# 아래에 코드를 작성하고 결과를 확인합니다.
# 아래에 코드를 작성하고 결과를 확인합니다.
print("유출이 제일 많은 구는?",seoul_moving.groupby('출발 시군구 코드')['출발 시군구 코드'].count().idxmax())
print("유입이 제일 많은 구는?",seoul_moving.groupby('도착 시군구 코드')['도착 시군구 코드'].count().idxmax())
print("유출이 제일 적은 구는?",seoul_moving.groupby('출발 시군구 코드')['출발 시군구 코드'].count().idxmin())
print("유입이 제일 적은 구는?",seoul_moving.groupby('도착 시군구 코드')['도착 시군구 코드'].count().idxmin())

유출이 제일 많은 구는? 11230
유입이 제일 많은 구는? 11230
유출이 제일 적은 구는? 11180
유입이 제일 적은 구는? 11180


In [22]:
# 아래에 코드를 작성하고 결과를 확인합니다.

seoul_moving.groupby('도착 시군구 코드')['이동인구(합)'].count().idxmax()

11230

#### [2.2.7] 낮시간만 분석 대상에 넣기 위해 심야버스 제외

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.

seoul_moving=seoul_moving.loc[~(seoul_moving['도착시간'].between(0,4))]
seoul_moving['도착시간'].values

seoul_moving.info()

seoul_moving['이동인구(합)']=seoul_moving['이동인구(합)'].astype('float64')

seoul_moving.info()

#### [2.2.8] '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)' , '이동인구(합)' 의 평균과 합 도출

* groupby 사용, as_index=False로 설정해주세요

In [ ]:
# '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)', '이동인구(합)'의 평균을 도출하여 seoul_moving_mean에 저장
seoul_moving_mean=seoul_moving.groupby('도착 시군구 코드',as_index=False)[['평균 이동 시간(분)','이동인구(합)']].mean()
seoul_moving_mean

In [ ]:
# '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)', '이동인구(합)'의 합을 도출하여 seoul_moving_sum에 저장

# '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)', '이동인구(합)'의 평균을 도출하여 seoul_moving_mean에 저장
seoul_moving_mean=seoul_moving.groupby('도착 시군구 코드',as_index=False)[['평균 이동 시간(분)','이동인구(합)']].mean()
seoul_moving_mean

In [ ]:
# 필요에 따라 column명을 데이터 의미에 맞게 적절히 변경하세요.
# 필요에 따라 column명을 데이터 의미에 맞게 적절히 변경하세요.
seoul_moving_mean=seoul_moving_mean.rename(columns={'평균 이동 시간(분)':'평균 이동 시간(분) 평균','이동인구(합)':'이동인구(합) 평균'})
seoul_moving_sum=seoul_moving_sum.rename(columns={'평균 이동 시간(분)':'평균 이동 시간(분) 총합','이동인구(합)':'이동인구(합) 총합'})

In [ ]:
# seoul_moving_mean, seoul_moving_sum을 합쳐보세요.
# seoul_moving_mean, seoul_moving_sum을 합쳐보세요.
df_seoul_moving=pd.merge(seoul_moving_mean,seoul_moving_sum)
df_seoul_moving

In [ ]:
# 해당 데이터프레임을 'df_seoul_moving.csv' 파일로 저장하세요


---

# 2.3.데이터 분석하기
+ 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별하세요.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# 시각화 한글폰트 설정
# 맑은 고딕

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

# scipy
import scipy.stats as spst

#### [2.3.1] 데이터 분포 알아보기
* 다양한 변수를 기준으로 그래프를 그려보고 인사이트를 도출해보세요.

In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.


In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.

In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.

In [ ]:
# 위 차트를 통해 알게된 사실을 정리해봅시다.
# 1. 
# 2.
# 3.

---